### PR TITLE
catch '${x.toString()}' with noop_primitive_operations

### PR DESCRIPTION
**Pull request by a14n**
# Description

Improve` noop_primitive_operations` to catch `.toString()` in string interpolations.

Related to #3334

### CODE REVIEW

**Review by bwilkerson — APPROVED** [review]
The implementation is good, but again this is a breaking change for anyone using it. I think it's worth doing, but might require some additional work to land.

Also, we don't have a fix for this lint, but this case, at least, is an easy fix. (That doesn't need to block merging this PR.)

**Comment by pq** [comment]
> The implementation is good, but again this is a breaking change for anyone using it.

This lint is enabled for (at least) a project internally so yeah, we'll have to see if any issues crop up.

> Also, we don't have a fix for this lint, but this case, at least, is an easy fix. 

Which is just to say, it'd be great to get this done before we publish a new linter with this improvement. 


PS: This looks great!

**Review by pq — APPROVED** [review]
👍

**Comment by a14n** [comment]
Not sure if I should have created another PR but I added the case of `print(xxx.toString())`

Let me know if you prefer another PR (I understood you were not going to merge it quickly so I made this  change here to avoid too much concurrent pending branchs)

**Comment by a14n** [comment]
I will add the corresponding fixes.

File: lib/src/rules/noop_primitive_operations.dart [hunk 1]
```diff
@@ -66,10 +69,27 @@ class _Visitor extends SimpleAstVisitor<void> {
...
+  }
+
   @override
   void visitMethodInvocation(MethodInvocation node) {
     var type = node.realTarget?.staticType;
-    if (type == null) return;
+    if (type == null) {
+      // print(xxx.toString())
+      if (node.methodName.name == 'print' &&
+          node.methodName.staticElement?.library?.name == 'dart.core') {
```
Comment (pq) [inline lib/src/rules/noop_primitive_operations.dart:88]: I wonder if we shouldn't prefer:

```dart
node.methodName.staticElement?.library?.isDartCore ?? false
```

(and if so, update `AvoidPrint` similarly).

Dunno.

@scheglov: do you all have comventions in analyzer?

**Comment by pq** [comment]
> Not sure if I should have created another PR but I added the case of print(xxx.toString())

Nah.  Here is fine.

And thanks!

**Comment by pq** [comment]
> I will add the corresponding fixes.

Thank you!

**Comment by a14n** [comment]
> I will add the corresponding fixes.

Actually I don't find existing corrections for diagnostics `noop_primitive_operations`. (I tested on `1.2.toDouble()`)

File: lib/src/rules/noop_primitive_operations.dart (hunk 1 above)
Comment (scheglov) [reply lib/src/rules/noop_primitive_operations.dart:88]: `isDartCore` is definitely better.

Also, I would have added an extension like `isDartCorePrint` on `Element?` and checked that it is `FunctionElement` with the required name and library.

**Comment by pq** [comment]
> Actually I don't find existing corrections for diagnostics noop_primitive_operations. (I tested on 1.2.toDouble())

Hmmmm.  Right.  It looks like that hasn't been added.  Looking in `error_fix_status.yaml` I see:

```yaml
    LintCode.noop_primitive_operations:
      status: needsEvaluation
```

File: lib/src/rules/noop_primitive_operations.dart (hunk 1 above)
Comment (pq) [reply lib/src/rules/noop_primitive_operations.dart:88]: Playing around with that now.  Stay tuned! 😄

Comment (pq) [reply lib/src/rules/noop_primitive_operations.dart:88]: Landed.  Thanks @scheglov!

@a14n: could you migrate to using `isDartCorePrint` as introduced by https://github.com/dart-lang/linter/pull/3382?

File: lib/src/rules/noop_primitive_operations.dart [hunk 2]
```diff
@@ -66,10 +70,21 @@ class _Visitor extends SimpleAstVisitor<void> {
...
+    _checkToStringInvocation(node.expression);
+  }
+
   @override
   void visitMethodInvocation(MethodInvocation node) {
     var type = node.realTarget?.staticType;
-    if (type == null) return;
+    if (type == null) {
+      // print(xxx.toString())
+      if (node.methodName.staticElement?.isDartCorePrint ?? false) {
```
Comment (scheglov) [inline lib/src/rules/noop_primitive_operations.dart:83]: You don't need `?.` here - the extension already works on `Element?`.

Which itself might be a good lint? :-)

File: lib/src/rules/noop_primitive_operations.dart [hunk 3]
```diff
@@ -66,10 +70,21 @@ class _Visitor extends SimpleAstVisitor<void> {
...
+        _checkToStringInvocation(node.argumentList.arguments.first);
```
Comment (scheglov) [inline lib/src/rules/noop_primitive_operations.dart:84]: The code under analysis might be incomplete, and have `print();` without any arguments.

File: lib/src/rules/noop_primitive_operations.dart (hunk 2 above)
Comment (a14n) [reply lib/src/rules/noop_primitive_operations.dart]: Should it go into `invalid_null_aware_operator` ?

**Comment by a14n** [comment]
Flutter will be diagnostic free once https://github.com/flutter/flutter/pull/103226 has landed.

File: lib/src/rules/noop_primitive_operations.dart (hunk 2 above)
Comment (a14n) [reply lib/src/rules/noop_primitive_operations.dart]: https://github.com/dart-lang/sdk/pull/48980 is an attempt to cover this case.

**Comment by a14n** [comment]
ping

**Comment by pq** [comment]
Apologies, @a14n, I wasn't able to make progress this week and now out for 2 of vacation.  I'll pick it up when I'm back.  Sorry for the delay!

**Comment by a14n** [comment]
@pq Friendly reminder ;)

**Comment by pq** [comment]
Thanks for the nudge and apologies for the slowness.  I've just kicked off an internal run.

**Comment by pq** [comment]
Looks good.  Thanks again for your patience.  And the fix! 🎉

**Pull request merged**